File: workers/worker_data-processing/src/adapter/services/queue_service.py
import asyncio
import aio_pika
import json

class QueueService:
  def __init__(self):
    self._connection = None
    self._channel = None
    self._exchange = None

  async def start(
    self,
    extension_name: str,
    user: str,
    password: str,
    handle_message: callable,
    max_retries: int = 5,
    retry_delay: int = 5,
  ):
    attempt = 0
    while attempt < max_retries:
      try:
        self._connection = await aio_pika.connect_robust(
          f"amqp://{user}:{password}@rabbitmq/"
        )

        self._channel = await self._connection.channel()
        await self._channel.set_qos(prefetch_count=10)

        self._exchange = await self._channel.declare_exchange(
          "tasks-exchange",
          aio_pika.ExchangeType.TOPIC,
          durable=True,
        )
        queue_name = f"{extension_name}-queue"
        queue = await self._channel.declare_queue(
          queue_name,
          durable=True,
        )

        routing_key = f"tasks.{extension_name}.#"
        await queue.bind(self._exchange, routing_key)

        await queue.consume(handle_message)
        await asyncio.Future()
      except Exception as error:
        attempt += 1
        print(f"[!] Connection failed ({attempt}/{max_retries}): {error}", flush=True)
        if attempt >= max_retries:
            raise
        await asyncio.sleep(retry_delay)

  async def publish_result(
    self,
    payload: dict,
  ):
    routing_key = "tasks.results"
    await self._publish_message(
      payload=payload,
      routing_key=routing_key,
    )

  async def publish_error(
    self,
    payload: dict,
  ):
    routing_key = "tasks.failed"
    await self._publish_message(
      payload=payload,
      routing_key=routing_key,
    )

  async def _publish_message(
    self,
    payload: dict,
    routing_key: str,
  ):
    message = aio_pika.Message(
      body=json.dumps(payload).encode(),
      content_type="application/json",
      delivery_mode=aio_pika.DeliveryMode.PERSISTENT,
    )

    await self._exchange.publish(
      message,
      routing_key=routing_key,
    )

queue_service = QueueService()

File: services/service_notification/src/queue/queue.service.ts
import { Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import amqp, { ChannelModel, Channel, ConsumeMessage } from 'amqplib';
import { ConfigService } from '@nestjs/config';
import { plainToInstance } from 'class-transformer';
import { validateOrReject } from 'class-validator';
import { QueueMessageDto } from './dtos/queue-message.dto';

@Injectable()
export class QueueService implements OnModuleInit, OnModuleDestroy {
  private exchange = 'tasks-exchange';
  private routingKey = 'tasks';
  private notificationRoutingKey = 'notification';

  private connection?: ChannelModel;
  private channel?: Channel;
  private isReconnecting = false;

  private handlers?: ((data: QueueMessageDto) => void)[];

  constructor(private readonly configService: ConfigService) {}

  async onModuleInit() {
    const queueUser = this.configService.getOrThrow<string>('rabbitmq.user');
    const queuePassword =
      this.configService.getOrThrow<string>('rabbitmq.password');
    await this.connectWithRetry(queueUser, queuePassword);
  }

  async onModuleDestroy() {
    await this.channel?.close();
  }

  /**
   * Set handlers for notification
   * @param handler handler for notification
   */
  setHandler(handler: (data: QueueMessageDto) => void) {
    if (!this.handlers) {
      this.handlers = [handler];
    } else {
      this.handlers = [...this.handlers, handler];
    }

    void this.startConsumers();
  }

  /**
   * Start result and failed task consumers
   */
  private async startConsumers() {
    if (!this.channel) return;
    if (!this.handlers) return;

    await this.consume(
      `${this.routingKey}.${this.notificationRoutingKey}`,
      'task-notification-queue',
    );
  }

  /**
   * Start consume from queue
   * @param routingKey Routing key for topic exchange
   * @param queueName Name of queue
   */
  private async consume(routingKey: string, queueName: string) {
    if (!this.channel) return;

    await this.channel.assertQueue(queueName, { durable: true });
    await this.channel.bindQueue(queueName, this.exchange, routingKey);

    await this.channel.consume(queueName, (message: ConsumeMessage | null) => {
      void (async () => {
        if (!message) return;

        try {
          const data: unknown = JSON.parse(message.content.toString());
          const dto = plainToInstance(QueueMessageDto, data);
          await validateOrReject(dto as object);

          for (const handler of this.handlers!) {
            handler(dto);
          }

          this.channel!.ack(message);
        } catch {
          this.channel!.nack(message, false, false);
        }
      })();
    });
  }

  /**
   * Connect to RabbitMQ with retry
   * @param user RabbitMQ user
   * @param password RabbitMQ password
   * @param maxRetries Count of retries attempts
   * @param retryDelay Retry delay in ms
   */
  private async connectWithRetry(
    user: string,
    password: string,
    maxRetries: number = 10,
    retryDelay: number = 3000,
  ) {
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        await this.connect(user, password);
        this.setupErrorHandlers(user, password);
        void this.startConsumers();
        return;
      } catch (error) {
        if (attempt === maxRetries) throw error;
        await new Promise((r) => setTimeout(r, retryDelay));
      }
    }
  }

  /**
   * Connect to RabbitMQ
   * @param user RabbitMQ user
   * @param password RabbitMQ password
   */
  private async connect(user: string, password: string): Promise<void> {
    const url = `amqp://${user}:${password}@rabbitmq:5672`;
    this.connection = await amqp.connect(url);
    this.channel = await this.connection.createChannel();

    await this.channel.assertExchange(this.exchange, 'topic', {
      durable: true,
    });
  }

  /**
   * Setup error handlers for reconnection
   * @param user RabbitMQ user
   * @param password RabbitMQ password
   */
  private setupErrorHandlers(user: string, password: string) {
    this.connection?.on('close', () => {
      void this.reconnect(user, password);
    });

    this.connection?.on('error', () => {
      void this.reconnect(user, password);
    });
  }

  /**
   * Reconnect to RabbitMQ
   * @param user RabbitMQ user
   * @param password RabbitMQ password
   */
  private async reconnect(user: string, password: string) {
    if (this.isReconnecting) return;

    this.isReconnecting = true;
    try {
      await this.connectWithRetry(user, password);
    } finally {
      this.isReconnecting = false;
    }
  }
}
